[Frontend] Node detail view now can show workflow input/output artifacts

eterna2 opened the pull request:
This is part 1 of the PR with reference to #2172 - to show workflow input/output artifacts in the front-end.

cc: @Ark-kun 

This PR:
- Add a new component `S3ArtifactLink` which renders a `<a href>` link (supports only s3:// and minio://)
- Updated `DetailsTable`:
  - add a new optional `valueComponent` prop: will wrap the value if provided 
  - props is changed from `string[][]` to `Array<[string?, any?]>` so that the value need not be a string.
- Updated `StaticNodeDetails` to include 2 additional fields: Input Artifacts and Output Artifacts.

The next PR will:
- update the node backend with a new get artifact API, with additional query params (namely the endpoint, and secrets from the workflow status)
- this endpoint should also support lookahead (i.e. if flag provided, will head the artifact size, and return either a link to the artifact or the actual content of the artifact if it is small)
- this endpoint should support artifacts with and w/o gzip
- `It should probably also support partial download with size/length (and maybe offset/start), so that it's possible to visualize the beginning of large table.` (@Ark-kun )

![Screen Shot 2019-10-04 at 7 49 51 PM](https://user-images.githubusercontent.com/1248825/66206725-84b61180-e6e3-11e9-9a48-640fe893c46e.png)

Video can be seen here:
https://drive.google.com/file/d/1GmNrJum20AdW3Bo0xDQJFIOvvMhCF7C_/view

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/kubeflow/pipelines/2305)
<!-- Reviewable:end -->

Ark-kun commented:
> * new get artifact API, with additional query params

It should probably also support partial download with `size`/`length` (and maybe `offset`/`start`), so that it's possible to visualize the beginning of large table.

Ark-kun commented:
There are some frontend unit test errors:

https://travis-ci.com/kubeflow/pipelines/jobs/242030777

eterna2 commented:
ok. updated the snapshot to match the current PR.

eterna2 commented:
```
It should probably also support partial download with size/length (and maybe offset/start), so that it's possible to visualize the beginning of large table.
```
@Ark-kun 

question. for ui-metadata it is not an issue, as the data type is specified. 

However, for generic artifacts, there is no easy way to know what is the content? other than try/catch with a few possible choices? But probably doable, do a quick scan for number of commas, or common delimiters to differeniate between csv, json, or plain text.

Ark-kun commented:
> However, for generic artifacts, there is no easy way to know what is the content?

This PR will make the type information available to Frontend: https://github.com/kubeflow/pipelines/pull/2323

>But probably doable, do a quick scan for number of commas, or common delimiters to differeniate between csv, json, or plain text.

I think that's out of scope for this PR. Here we just want to restore the existing functionality. Initially we just want to show short strings. We can truncate the string to, say, 300 bytes and show something like `"very long strin...", Size:  32167bytes, URL: https://....`.

Ark-kun commented:
/ok-to-test

Ark-kun commented:
> * `It should probably also support partial download with size/length (and maybe offset/start), so that it's possible to visualize the beginning of large table.` (@Ark-kun )

Note: I was talking about the Frontend server functionality. This partial download feature should not be exposed to the user.

Ark-kun commented:
Are the artifact links in "clickable" - result in showing/downloading the artifact data?

eterna2 commented:
Yup. It opens a new window and shows the data in plain text. I am just using the original get artifact API in the node backend as a HREF link in this PR (just a new window with a GET request).

However to trigger download instead of a new window, I need to add a `Content-Disposition` header to the API response.

Probably, can add a new query string flag `download` to set this.

Bobgy commented:
I'll review this today.

Ark-kun commented:
> However to trigger download instead of a new window, I need to add a `Content-Disposition` header to the API response.
> 
> Probably, can add a new query string flag `download` to set this.

It's OK. People can always right-click the URL and use "Save as".

Ark-kun commented:
> I'll review this today.

Thank you.

Bobgy commented on frontend/src/components/S3ArtifactLink.tsx:
```diff
@@ -0,0 +1,31 @@
+  s3artifact
+) => {
+  if (!s3artifact || !s3artifact.key || !s3artifact.bucket) {
+    return <>{}</>;
```
nit: `return null` instead?

eterna2 replied:
fixed.

Bobgy commented on frontend/src/lib/StaticGraphParser.ts:
```diff
@@ -68,7 +68,7 @@ export function _populateInfoFromTemplate(info: SelectedNodeInfo, template?: Tem
   if (template.inputs) {
     info.inputs =
-      (template.inputs.parameters || []).map(p => [p.name, p.value || '']);
+      (template.inputs.parameters || []).map(p => [p.name, p.value || ''] as [string, string]);
```
nit: Can you give [string, string] a proper `type/interface` name and avoid casting here? Avoiding type casting helps with long term maintenance.

Bobgy replied:
One way to avoid casting here is
```
....map((p): [string, string] => [p.name, p.value || '']);
```
[example](http://www.typescriptlang.org/play/?ts=3.1.6#code/MYewdgzgLgBAhgJwQLhgQSXAngHgNrQICWYA5gDQyEmkC6AfDALwBQM7MeAjLQHQC2cAA4AKEQEpkBKMTKVqZWs0Z4A5HABGwVZVUATAKYAzVbXEsgA)

What would be even better, we should probably upgrade typescript version to at least 3.5, so it can handle it by default. [example](http://www.typescriptlang.org/play/?ts=3.5.1#code/MYewdgzgLgBAhgJwQLhgQSXAngHgNrQICWYA5gDQyEmkC6AfDALwBQM7MeAjLQHQC2cAA4AKEQEpmjPAHI4AI2AzKMgCYBTAGYza4lkA)

eterna2 replied:
updated typescript to 3.6.

eterna2 replied:
fixed issues introduced by the stricter typing by typescript >=3.5.

updated node to LTS 10 - because I notice a few peculiarity:

- the grpc packages are not found in `package-lock.json` even thou it specified in `package.json` (I think is subsequently installed in the post-install step, but I am not sure if it is used in any other places) - so whenever I do a `npm i` they get included into the lock file. And this result in some issues for node 9.4, as some dependent packages seems to be missing (does not support node 9)? So I have to upgrade the docker images to node 10.

- Some of packages are not following the semantic versioning faithfully - I have to freeze a couple of packages, as the latest ones breaks the app (when `package-lock.json` is updated when I upgraded `typescript`.

Bobgy replied:
> the grpc packages are not found in package-lock.json even thou it specified in package.json (I think is subsequently installed in the post-install step, but I am not sure if it is used in any other places) - so whenever I do a npm i they get included into the lock file. And this result in some issues for node 9.4, as some dependent packages seems to be missing (does not support node 9)? So I have to upgrade the docker images to node 10.

Thanks for the catch. I think the last committer forgot to update package-lock.json last time, let's just commit them so they are in sync. Besides upgrading to node 10 LTS is definitely a good choice.

> Some of packages are not following the semantic versioning faithfully - I have to freeze a couple of packages, as the latest ones breaks the app (when package-lock.json is updated when I upgraded typescript.

Thanks for finding them, the solution sounds good to me.

Bobgy commented on frontend/src/components/S3ArtifactLink.tsx:
```diff
@@ -0,0 +1,31 @@
+/**
+ * A component that renders an artifact link.
+ */
+const S3ArtifactLink: React.FC<S3Artifact> = (
```
May I ask for a really simple test for this component?

eterna2 replied:
added 3 simple test case:
- invalid artifact (empty bucket or key)
- s3 artifact
- minio artifact

Bobgy replied:
Thanks! LGTM

Bobgy commented on frontend/src/components/S3ArtifactLink.tsx:
```diff
@@ -0,0 +1,31 @@
+import { StoragePath, StorageService } from '../lib/WorkflowParser';
+import { S3Artifact } from '../../third_party/argo-ui/argo_template';
+
+const artifactApiUri = ({ source, bucket, key }: StoragePath) =>
```
It looks like this works with minio backed by gcs/local disk too. Is my understanding correct?

If that's the case, shall we rename this component to `MinioArtifactLink`?

eterna2 replied:
yup. Have renamed to MinioArtifactLink.

Bobgy commented:
Also, for convenience, can you record a video about how this looks like?
There are a lot of tools to do that, maybe you can try https://www.screencastify.com/

eterna2 commented:
@Bobgy 

List of changes:
- Upgraded typescript to 3.6 and fixed issues due to stricter type checking, added config to skip typecheck for material lib (as it breaks with new ts)
- Upgraded docker image to use node 10 instead of node 9, because grpc is broken in node 9
- freezed the version for a couple of packages because the latest in semantic version breaks the typechecks
- add apache-2 LICENSE (his github used apache-2, but npm show ISC) for `linear-layout-vector` (seems to be a new dependency for `react-virtualized`) 

video can be found at https://drive.google.com/file/d/1GmNrJum20AdW3Bo0xDQJFIOvvMhCF7C_/view

Ark-kun commented:
/retest

Bobgy commented:
Hi @eterna2, this is awesome! Thanks a lot for the quick change.

I hope I said this earlier because I didn't expect you to finish the upgrade so fast (which is awesome), but can you separate the following to a separate PR?
> * Upgraded typescript to 3.6 and fixed issues due to stricter type checking, added config to skip typecheck for material lib (as it breaks with new ts)
> * Upgraded docker image to use node 10 instead of node 9, because grpc is broken in node 9
> * freezed the version for a couple of packages because the latest in semantic version breaks the typechecks
> * add apache-2 LICENSE (his github used apache-2, but npm show ISC) for linear-layout-vector (seems to be a new dependency for react-virtualized)

Because this PR is getting too large and upgrading typescript brought too many side effects.

I would recommend finishing this PR without upgrading typescript (just use the pattern I introduced to avoid type casts).
```
....map((p): [string, string] => [p.name, p.value || '']);
```

Feel free to talk about your opinions because I can see you've already put a lot of efforts into this. I'm just giving suggestions that can give us more confidence of merging each of your PRs, and maybe sometimes I'm asking too much.

eterna2 commented:
@Bobgy 

u can't do that with typescript 3.3, because it doesn't handles tuple well in a `map/filter/reduce` operation - ts 3.5 fixed that thou.

the type check error is not due to the optional string, and the original code solves this by putting the type as `string[][]` instead of a proper array of tuples (`Array<[string, string?]`).

I cannot use this workaround because I wanted a `Array<[string, any]>`, where the value is either a string or prop, which can be pass to a React component. 

The only solution for now is to create an interface.

Bobgy commented:
@eterna2 I'm not sure what I missed here, I think my [code example](http://www.typescriptlang.org/play/?ts=3.3.3#code/MYewdgzgLgBAhgJwQLhgQSXAngHgNrQICWYA5gDQyEmkC6AfDALwBQM7MeAjLQHQC2cAA4AKEQEpkBKMTKVqZWs0Z4A5HABGwVZVUATAKYAzVbXEsgA) shows a workaround to this problem -- explicitly annotate return type of map function to let it be understood as [string, any] instead of string[].

But nevermind on details, I think upgrade is the best option, we've had other issues before too with this typescript version.

Do you want to put together a PR just for upgrading typescript? If not, I can also migrate your work here and prioritize that PR getting merged first.

Let me know your thoughts.

eterna2 commented:
@Bobgy 

for typescript 3.3, tuples are not resolvable for `map/filter/reduce` operations, i.e

```ts
// this will throw a typecheck error as result will be string[][] instead of Array<[string, string?]>
var result: Array<[string, string?]> = someArrayOfTuples.map(([k, v]) => [k, v || '']) // this still resolves to array of array instead of array of tuples.
```
U need typescript >= 3.5 to resolve tuples properly after a `map` function.

Anyway, I have made the changes u requested:
- Added `null`, and `{}` to unit test
- reverted changes to typescript version (and all the patches)
- Created an interface `KeyValue` which inherits from `Array` so that type casting is no required for `map` operations.

Bobgy commented on frontend/src/lib/StaticGraphParser.ts:
```diff
@@ -56,7 +61,7 @@ export function _populateInfoFromTemplate(info: SelectedNodeInfo, template?: Tem
     info.command = template.container.command || [],
     info.image = template.container.image || '';
-    info.volumeMounts = (template.container.volumeMounts || []).map(v => [v.mountPath, v.name]);
+    info.volumeMounts = (template.container.volumeMounts || []).map(v => [v.mountPath, v.name] );
```
nit: can you revert these extra spaces?

Bobgy commented on frontend/src/lib/StaticGraphParser.ts:
```diff
@@ -78,7 +83,7 @@ export function _populateInfoFromTemplate(info: SelectedNodeInfo, template?: Tem
         value = p.valueFrom.jqFilter || p.valueFrom.jsonPath || p.valueFrom.parameter || p.valueFrom.path || '';
       }
-      return [p.name, value];
+      return [p.name, value] ;
```
also here

Bobgy commented on frontend/src/lib/WorkflowParser.ts:
```diff
@@ -168,49 +169,69 @@ export default class WorkflowParser {
     const inputsOutputs: [paramList, paramList] = [[], []];
     if (node.inputs && node.inputs.parameters) {
-      inputsOutputs[0] = node.inputs.parameters.map(p => [p.name, p.value || '']);
+      inputsOutputs[0] = node.inputs.parameters.map(p => [p.name, p.value || ''] );
```
and here

Bobgy commented on frontend/src/lib/WorkflowParser.ts:
```diff
@@ -168,49 +169,69 @@ export default class WorkflowParser {
     }
     if (node.outputs && node.outputs.parameters) {
-      inputsOutputs[1] = node.outputs.parameters.map(p => [p.name, p.value || '']);
+      inputsOutputs[1] = node.outputs.parameters.map(p => [p.name, p.value || ''] );
```
and here

Bobgy commented on frontend/src/lib/WorkflowParser.ts:
```diff
@@ -168,49 +169,69 @@ export default class WorkflowParser {
-  public static getNodeInputOutputParams(workflow?: Workflow, nodeId?: string): [string[][], string[][]] {
-    type paramList = string[][];
+  public static getNodeInputOutputParams(workflow?: Workflow, nodeId?: string): [Array<KeyValue<string>>, Array<KeyValue<string>>] {
+    type paramList = Array<KeyValue<string>>;
```
nit: naming convention, type always start with uppercase `ParamList`

Bobgy commented on frontend/src/pages/RunDetails.tsx:
```diff
@@ -377,7 +387,7 @@ class RunDetails extends Page<RunDetailsProps, RunDetailsState> {
                   {workflowParameters && !!workflowParameters.length && (<div>
                     <DetailsTable title='Run parameters'
-                      fields={workflowParameters.map(p => [p.name, p.value || ''])} />
+                      fields={workflowParameters.map(p => [p.name, p.value || ''] )} />
```
and here

Bobgy commented on frontend/src/pages/RecurringRunDetails.tsx:
```diff
@@ -65,15 +66,15 @@ class RecurringRunDetails extends Page<{}, RecurringRunConfigState> {
         ['Created at', formatDateString(run.created_at)],
       ];
-      inputParameters = (run.pipeline_spec.parameters || []).map(p => [p.name || '', p.value || '']);
+      inputParameters = (run.pipeline_spec.parameters || []).map(p => [p.name || '', p.value || ''] );
```
and here

Bobgy commented on frontend/src/lib/WorkflowParser.ts:
```diff
@@ -168,49 +169,69 @@ export default class WorkflowParser {
+  // Makes sure the workflow object contains the node and returns its
+  // inputs/outputs artifacts if any, while looking out for any missing link in the chain to
+  // the node's inputs/outputs.
+  public static getNodeInputOutputArtifacts(workflow?: Workflow, nodeId?: string): [Array<KeyValue<S3Artifact>>, Array<KeyValue<S3Artifact>>] {
```
nit: an object return value like `{ inputArtifacts, outputArtifacts }` clarifies both array contents better.

I can see you are following the same convention as `getNodeInputOutputParams` which is not a good example by itself.

Bobgy replied:
Also, please add a test similar to this one: https://github.com/kubeflow/pipelines/blob/cd0aeb6c6297de8dd58a979cd172cfb611bfa730/frontend/src/lib/WorkflowParser.test.ts#L339

eterna2 replied:
do u mean to change the return value from a tuple to an object?
or do u mean a tuple `[inputArtifacts, outputArtifacts]` ? with a type alias?

cuz changing the return value means a lot of changes downstream.

Bobgy replied:
Yes, I think changing the return value to an object makes usage code more readable. This has only one usage, so I believe it's fine.

eterna2 replied:
ok, done.
I also made the changes for `getNodeInputOutputParams`.
Updated the tests for both of them too.

Bobgy commented on frontend/src/components/DetailsTable.tsx:
```diff
@@ -40,43 +42,60 @@ export const css = stylesheet({
-  fields: string[][];
+  fields: Array<KeyValue<string|S3Artifact>>;
   title?: string;
+  valueComponent?: React.FC<S3Artifact>;
```
Can you add a simple test for this field in `DetailsTable.test.tsx`?

eterna2 replied:
ok done.

Bobgy commented:
/retest

eterna2 commented:
ok. fixed most of the minor stuff.

wonders whether do u want to setup an auto-formatter like `prettier`? Cuz I usually does that, and just auto-format everything as a pre-hook commit.

will fix the unit test and output later.

Bobgy commented:
Thanks for suggestion, I am also planning on making a standard prettier config.

eterna2 commented on frontend/src/components/MinioArtifactLink.test.tsx:
```diff
@@ -0,0 +1,54 @@
+      key: '',
+      secretKeySecret: {key: 'secretkey', optional: false, name: 'minio'},
+    };
+    expect(MinioArtifactLink(s3artifact)).toMatchSnapshot();
```
done.

eterna2 commented:
@Bobgy 
Ok done with all the changes if I am not wrong. 

Just a minor feedback, the unit test fails on node 9 (travis is running node 11, while dockerfile is using node 9) - might want to sync them up, cuz was puzzled by my failing unit tests until I went and look at travis and swapped to 11 (the errors are all datetime format diff with the snapshots).

But this can be specific to my machine thou. I am using wsl in windows.

Ark-kun commented:
> (travis is running node 11, while dockerfile is using node 9) - might want to sync them up

@Bobgy Which version should we standardize on: 9 ,10, 11 or 12? The NodeJS website lists 10 and 12 as LTS releases. Should we try to go with 12?

Bobgy commented:
@Ark-kun Choosing latest LTS (12) is definitely the best option. I agree we should do it in a separate PR. I will create an issue for it.

Bobgy commented:
@eterna2, thanks so much for this! This is awesome.

/lgtm
/approve

Bobgy commented:
/area front-end
